### ISSUE TITLE
Yanking policy clarification/discussion

### ISSUE THREAD

**Issue by djc** [google/zerocopy#679]
My project at work was affected by the yanking of 0.7.28 yesterday. I was curious about the reason (which wasn't trivial to find, since there don't seem to be any release notes) and ran into your new policy from #677:

> Whenever a bug or regression is identified, we will yank any affected versions
> which are part of the current version train. For example, if the most recent
> version is 0.10.20 and a bug is uncovered, we will release a fix in 0.10.21 and
> yank all 0.10.X versions which are affected. We *may* also yank versions in previous
> version trains on a case-by-case basis, but we don't guarantee it.

This doesn't have any language on whether there is a severity bar that a bug must meet before yanking will be done. IME yanking can be painful to downstream consumers (for example, we manage our lock files in version control and update them generally once a week, but use cargo-deny check to check for yanked crates, so this caused our CI to fail). It looks like the offending issue (in #672) was pretty serious, but it would be a pain if yanking started happening frequently (more than say once every few months).

**Comment by ijackson** [comment]
Reading #672 it's not clear to me if it was a soundness bug.  If it was, there probably should be a RustSec advisory?

**Comment by jswrenn** [comment]
@ijackson Not a soundness bug, fortunately! Just a completeness bug. The bug reported in #672 describes a situation in which sound code that was previously accepted was no longer accepted. 

Where things get a bit confusing is that we didn't think our `repr` parsing code should have ever handled this (sound) use of `packed(N)`; an implementation limitation we believed existed did not, in fact, exist. Fortunately, we have redundancy in our padding checks — we don't just look at the presence (or lack) of `repr(packed)` — and so there was no soundness issue.

@djc, @joshlf  Hm, perhaps we can take a more moderate policy of only yanking versions with soundness issues.

**Comment by ijackson** [comment]
> @djc, @joshlf Hm, perhaps we can take a more moderate policy of only yanking versions with soundness issues.

I think that would be a better policy.  It's what's usually done in other crates.

Yanking a version with a compilation regression doesn't really help anyone (assuming you're also publishing a fixed version).  Yanking means people who aren't affected by the bug might find that their lockfile breaks.  And as for people who *are* affected by the bug: if their lockfile has the broken version, then this just changes the failure mode but doesn't actually fix anything - those people have to update their lockfile anyway.

(People who don't maintain an in-tree lockfile will typically get the fixed version whether or not you yank the broken one.)

Thanks for your attention.

**Comment by joshlf** [comment]
Thanks for filing this! We don't have any visibility into how our decisions affect our users without issues like these, so it's really helpful.

@jswrenn and I discussed this and came up with the following proposal for a new yanking policy. How does this sound to y'all?

This table is read top-to-bottom. As soon as you encounter a row that applies, you stop and yank or don't yank as prescribed by the right-hand column.

| Condition | Yank? |
|-----------|-------|
| The bug causes compilation to fail | Don't yank |
| The bug affects soundness | Yank |
| The bug contradicts documented behavior | Yank |
| The bug contradicts behavior which is widely assumed | Yank |
| The bug is a regression (code exists which would succeed on a version X and would fail on a version Y > X) | Yank |
| Otherwise | Don't yank |

**Comment by ijackson** [comment]
Tools like `cargo audit` report on yanked crates as well as advisories.  So you get a report that the crate was yanked, and CI failure reports (depend on your precise workflow and CI setup).  You don't get any references to what is wrong.

But usually the Rust community don't yank things unless it's an emergency.  So the signal you get is "CI says this crate is on fire!" but without any extra info.

Also that same community norm can mean that "yanked crate" generates more intrusive failures/alerts/whatever.  Following that norm is helpful for your downstreams.

From the point of view of a downstream, being told "the zerocopy you shipped to people was yanked, now dig through upstream tickets to find out what the crisis is" is just a much worse experience than "zerocopy affected by RUSTSEC-2023-XXXYYY"

**Comment by joshlf** [comment]
I see, that's helpful context, thanks! If we adopt a policy of always publishing a rustsec advisory and [keeping a yank log in a `CHANGELOG.md` file](https://github.com/google/zerocopy/issues/238) in addition to yanking, would that mitigate these concerns?

**Comment by ijackson** [comment]
Certainly, yes.  The advisory is the most important since that's a formal structured database that's the same for every crate.  RUSTSEC keep advisories for yanked crates, obviously, so it would show up.

Thanks.

**Comment by ijackson** [comment]
(TBH I still don't see that there's much value in yanking; it's quite a crude mechanism.  But I think your suggestions would mitigate or eliminate the main downsides.)

**Comment by joshlf** [comment]
I've put up a policy in #722; I'll wait until tomorrow to merge it so folks have a chance to provide feedback if they want.

**Comment by djc** [comment]
My OP did mention that we're using cargo-deny, and we opted-in to failing CI on yanked dependencies. This is only tenable, of course, if the ecosystem norms align on yanking being reserved for fairly serious issues. IIRC the *ring* crate used to yank all releases that had bugs and there was quite a bit of backlash around that.

There's a [long-standing Cargo issue](https://github.com/rust-lang/cargo/issues/2608) about making it possible to submit a note when yanking, which could be used to clarify.

IMO just getting a new (semver-compatible) version out it is the right thing for all but the most serious (unsoundness, vulnerabilities, "widely assumed" etc) issues. As I understand it, semver-compatible releases get picked up pretty quickly due to Cargo's default behavior to pick the latest.

**Comment by joshlf** [comment]
> My OP did mention that we're using cargo-deny, and we opted-in to failing CI on yanked dependencies. This is only tenable, of course, if the ecosystem norms align on yanking being reserved for fairly serious issues. IIRC the _ring_ crate used to yank all releases that had bugs and there was quite a bit of backlash around that.

Yeah I definitely agree that frequent yanking would be problematic. We generally move very slowly and deliberately, and so historically have had very few user-facing bugs, so my expectation is that, despite this proposed policy, we would yank very infrequently. We have only recently begun to track yanking closely, so this is a rough estimation from memory, but I believe we've yanked on fewer than five occasions in zerocopy's five years of being on crates.io.

> IMO just getting a new (semver-compatible) version out it is the right thing for all but the most serious (unsoundness, vulnerabilities, "widely assumed" etc) issues. As I understand it, semver-compatible releases get picked up pretty quickly due to Cargo's default behavior to pick the latest.

My comment [here](https://github.com/google/zerocopy/pull/722#discussion_r1427054426) covers this, but the TLDR is that our perspective is that we need to be as conservative as our most conservative user since we don't have any visibility into which users would care about any given bug. While I agree that there are users who wouldn't care about some of the cases in which we're planning to yank, we'd prefer to slightly inconvenience these users than permit known bugs to remain in the codebases of our users who have a higher bar for correctness.

**Comment by djc** [comment]
> Yeah I definitely agree that frequent yanking would be problematic. We generally move very slowly and deliberately, and so historically have had very few user-facing bugs, so my expectation is that, despite this proposed policy, we would yank very infrequently. We have only recently begun to track yanking closely, so this is a rough estimation from memory, but I believe we've yanked on fewer than five occasions in zerocopy's five years of being on crates.io.

Okay, that doesn't sound too bad then.
